Scavengers: Refactor and add new features

sparkychildcharlie opened the pull request:
General
-------
- use RoomGame API for scavengers
- added queuing of games
- added a timer command for scavenger hunts
- allowed the hunt's host to view their hints/answers, and to edit them after the hunt has started
- display the hunt to users joining the room, if they're not already a player in the hunt
- checks IP and prevents users from joining and finishing a hunt several times simply by changing their userid
- support more than 3 questions for scavengers for potential events (approved [but not requested] by scavenger roomauth, although still with some debate)

Requested
---------
- split help command into 2 different sections that can be called independently 
- split the old /scavengerstatus into /scavengerstatus and /scavengerhint

Refactored
----------
- (Entire game system with RoomGame and RoomGamePlayer)
    - Blitzes will now give points regardless if the user finishes top 5, as specified in the [Scavengers Guide](https://docs.google.com/document/d/1Uk3n0iPNXuNJZ5zXdpTqa-gSMaTvX_pxKeYA9As2Xuk/edit):
    >"Officials also have a special “BLITZ” mechanic, which will be given any finisher if they finish within a minute: they get 10 bonus points."
- Ladder
    - refactored ladder function to be async to prevent any abuse of the feature (say 500 bots spamming it could potentially cause some I/O blocking)
    - bold only the top 3 non roomauth (requested by staff)
    - better checking for potential constructor abuse in ladder commands
    - (retain database compatibility/format with the pre-existing leaderboard)
    - update name display in the ladder only when points from from a WIN

*For reference purposes this has been directly approved by scavenger RO Flerovium (received blessing to pull request this), as well as comments/criticism and indirect approval from GME, and some of the moderators of the aforementioned room.

Zarel commented:
Can you find Scavengers staff and get them to approve it? A Smogon group PM would probably be the simplest approach.

panpawn commented on chat-plugins/scavengers.js:
```diff
@@ -1,303 +1,603 @@
-* This game shows a hint. Once a player submits the correct answer, they move on to the next hint.
-* You finish upon correctly answering the third hint.
-* In an official hunt, the first three to finish within 60 seconds achieve blitz.
-*/
```
I don't think we need to remove this documentation, unless this description is no longer valid?

sparkychildcharlie replied:
The line about the blitz isn't valid.

If something __were__ to be put there, wouldn't the MIT stuff usually make more sense, since that's whats on most of the other files.

panpawn replied:
Yes

panpawn commented on chat-plugins/scavengers.js:
```diff
@@ -1,303 +1,603 @@
-		scavengersRoom.plugin = scavengers;
-	}
-}
+const fs = require("fs");
```
tic marks

panpawn commented on chat-plugins/scavengers.js:
```diff
@@ -1,303 +1,603 @@
-}
+const fs = require("fs");
+
+const dataFile = "./config/chat-plugins/scavdata.json";
```
tic marks

panpawn commented on chat-plugins/scavengers.js:
```diff
@@ -1,303 +1,603 @@
-		scavengersLB = {};
-		return;
-	}
+	if (err) return console.log("ERROR: Unable to load scavenger leaderboard: " + err);
```
template string

panpawn commented on chat-plugins/scavengers.js:
```diff
@@ -1,303 +1,603 @@
-function getLadder() {
-	return Object.keys(scavengersLB).map(id => [scavengersLB[id].points, scavengersLB[id].name]).sort((a, b) => b[0] - a[0]);
+function visualizeLadder(userid) {
+	userid = toId(userid);
```
??

don't pass something to a function as ``userid`` if it isn't already a userid

panpawn commented on chat-plugins/scavengers.js:
@Zarel - I thought you once said multi-line template strings aren't a good idea, but I'm not entirely sure now?

sparkychildcharlie commented on chat-plugins/scavengers.js:
IIRC, back during the UNO stuff, I was told to split some super long template strings into several smaller chunks and then "concatenate" them

sparkychildcharlie commented:
@Zarel technically I have [this](http://prntscr.com/ep1ip1), but if you want me to get that to happen on Smogon as well, I guess I could do that as well...

sparkychildcharlie commented:
I have all the stuff changed locally, i'll wait for @Zarel 's input regarding the blob at the top of the plugin as well as any other comments before I push the second commit.

sparkychildcharlie commented:
- fixed the strings I've overlooked - I'm way too used to using ``"str" + "str2"`` for coding
- added the blurb at the top, modeled after the one in the room-events/uno plugin.

Zarel commented on chat-plugins/scavengers.js:
This isn't a multi-line template string.

```js
`this
 is a multi line template string; we do not want them in PS`
```

```js
`this` +
` is fine, on the other hand`
```

panpawn commented:
> [16:37] Dylas: I was reading the thing sparky put up for the code and I saw you needed approval
[16:38] Dylas: well, Zarel did
[16:38] %panpawn: scavengers?
[16:38] Dylas: yea
[16:38] %panpawn: I think he meant from the ROs
[16:38] Dylas: yes
[16:38] Dylas: I made a convo with all 3 ros, and all 3 gave approval

(Dylas appears to be a room mod in Scavengers, so I assume that's a relatively reliable source)

He also just showed me this:
<img src="http://image.prntscr.com/image/77978991f37c4ba6a3cfd0eea895422d.png">

Zarel commented:
Why wouldn't he add me to the convo?????

sparkychildcharlie commented:
bump @Zarel is this ready yet, since there hasn't been any comments/constructive criticism for 3 days already

Zarel commented on chat-plugins/scavengers.js:
```diff
@@ -1,303 +1,612 @@
-	}
-}
+const dataFile = './config/chat-plugins/scavdata.json';
+const points = [20, 15, 10, 5, 1];
```
I would probably capitalize these two constants.

Zarel commented on chat-plugins/scavengers.js:
```diff
@@ -1,303 +1,612 @@
-		};
-	}
+
+function addPoints(name, points, noUpdate) {
```
Free-floating functions look weird to me in general. Perhaps this should be a static function in your scavenger hunt class?

Or perhaps you should have a separate leaderboard class?

sparkychildcharlie replied:
so you would want me to make a "database" class?

so something like
```

class Database {
    constructor (path) {
          this.db = parse(readfile(path));
    }
    addPoints() {}
    write() {}
    visualizeLadder() {}
}
```

sparkychildcharlie replied:
Hm... maybe static function would be best.  Do you want the ladder function to be a static function in the scavhunt class as well?

Zarel replied:
A leaderboard class would make more sense, I think, but either works.

sparkychildcharlie commented:
- Moved all the ladder/leaderboard functions into a "self-contained" ``class`` created when ``scavengers.js`` is loaded.
- Capitalized the mentioned ``const`` variables

Zarel commented on chat-plugins/scavengers.js:
```diff
@@ -1,303 +1,629 @@
-		scavengersRoom.plugin = scavengers;
+const Points = [20, 15, 10, 5, 1];
+const path = require('path');
+const DataFile = path.resolve(__dirname, '../config/chat-plugins/scavdata.json');
```
By "capitalize", I mean that these constants would be named `POINTS` and `DATA_FILE`.

Like in https://github.com/Zarel/Pokemon-Showdown/blob/82ff7327fd3590d8903a586019df5a0fe30b2ab0/rooms.js#L14

sparkychildcharlie commented:
- Made those 2 variables entirely upper case

panpawn commented on chat-plugins/scavengers.js:
```diff
@@ -1,303 +1,629 @@
-		return;
+	load() {
+		fs.readFile(this.file, 'utf8', (err, content) => {
+			if (err) return console.log(`ERROR: Unable to load scavenger leaderboard: ${err}`);
```
... Does this not automatically make the data file?

panpawn replied:
You should automatically make the data file.

panpawn replied:
Never mind, it does do it automatically if/when it has data to write, my bad
